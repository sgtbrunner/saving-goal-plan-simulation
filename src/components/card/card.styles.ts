import styled from 'styled-components';

import { SMALL_DEVICE_BREAKPOINT } from '../../styles/global.styles';

export const CardContainer = styled.div`
  display: flex;
  flex-direction: column;
  padding: 31px 24px 24px 24px;
  background-color: var(--white-color);
  border-radius: 8px;
  margin-bottom: 63px;
  width: 100%;
  box-shadow: 0px 16px 32px rgba(30, 42, 50, 0.08);

  @media (min-width: ${SMALL_DEVICE_BREAKPOINT}px) {
    padding: 35px 40px 40px 40px;
    font-size: var(--md-font-size);
    line-height: var(--md-line-height);
    margin-bottom: 96px;
    width: auto;
  }
`;

export const CardTitle = styled.div`
  display: flex;
  align-items: center;

  img {
    padding-right: 24px;
  }

  h1,
  p {
    margin: 0;
  }

  h1 {
    font-weight: var(--normal-font-weight);
    font-size: var(--md-font-size);
    line-height: var(--md-line-height);
    color: var(--blue-gray-900-color);
    padding-bottom: 4px;
  }

  p {
    font-weight: var(--light-font-weight);
    font-size: var(--xxs-font-size);
    line-height: var(--xs-line-height);
    color: var(--blue-gray-400-color);
  }

  @media (min-width: ${SMALL_DEVICE_BREAKPOINT}px) {
    h1 {
      font-size: var(--lg-font-size);
      line-height: var(--lg-line-height);
    }

    p {
      font-size: var(--xs-font-size);
      line-height: var(--md-line-height);
    }
  }
`;

export const CardForm = styled.form`
  display: flex;
  flex-direction: column;

  label {
    padding-top: 16px;
    padding-bottom: 4px;
  }

  @media (min-width: ${SMALL_DEVICE_BREAKPOINT}px) {
    flex-direction: row;
    gap: 16px;

    label {
      padding-top: 24px;
    }
  }
`;
