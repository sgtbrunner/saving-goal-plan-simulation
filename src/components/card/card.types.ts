export type CardProps = {
  title: string;
  subtitle: string;
  icon: string;
  buttonText?: string;
};
