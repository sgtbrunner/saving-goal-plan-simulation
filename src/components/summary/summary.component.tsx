import { SummaryProps, FormatCurrencyType } from './summary.types';
import {
  SummaryContainer,
  MonthlyAmountContainer,
  SummaryTextContainer,
} from './summary.styles';

const USDCurrencyFormatter = (digits: number): Intl.NumberFormat =>
  new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    maximumFractionDigits: digits,
  });

const formatUSDCurrency = ({ value, digits = 2 }: FormatCurrencyType): string =>
  USDCurrencyFormatter(digits).format(value);

const Summary = ({
  goalAmount,
  reachDateMonth,
  reachDateYear,
  monthlyAmount,
  monthlyDeposits,
}: SummaryProps): JSX.Element => (
  <SummaryContainer>
    <MonthlyAmountContainer>
      <h4>Monthly amount</h4>
      <p>{formatUSDCurrency({ value: monthlyAmount })}</p>
    </MonthlyAmountContainer>
    <SummaryTextContainer>
      <p>
        Youâ€™re planning <strong>{`${monthlyDeposits} monthly deposits`}</strong>{' '}
        to reach your{' '}
        <strong>{formatUSDCurrency({ value: goalAmount, digits: 0 })}</strong>{' '}
        goal by{' '}
        <strong>
          {reachDateMonth} {reachDateYear}
        </strong>
        .
      </p>
    </SummaryTextContainer>
  </SummaryContainer>
);

export default Summary;
