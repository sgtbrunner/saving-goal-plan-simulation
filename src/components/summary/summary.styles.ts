import styled from 'styled-components';

import { SMALL_DEVICE_BREAKPOINT } from '../../styles/global.styles';

export const SummaryContainer = styled.div`
  display: flex;
  flex-direction: column;
  border: 1px solid var(--blue-gray-50-color);
  border-radius: 8px;
  margin: 24px 0 32px 0;
  width: 100%;

  @media (min-width: ${SMALL_DEVICE_BREAKPOINT}px) {
    max-width: 490px;
  }
`;

export const MonthlyAmountContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 32px 32px 24px 32px;

  h4 {
    margin: 0;
    font-size: var(--sm-font-size);
    line-height: var(--sm-line-height);
    font-weight: var(--light-font-weight);
    color: var(--blue-gray-900-color);
  }

  p {
    margin: 0;
    font-size: var(--lg-font-size);
    line-height: var(--lg-line-height);
    font-weight: var(--normal-font-weight);
    color: var(--secondary-color);
  }

  @media (min-width: ${SMALL_DEVICE_BREAKPOINT}px) {
    h4 {
      font-size: var(--md-font-size);
      line-height: var(--md-line-height);
    }

    p {
      font-size: var(--xlg-font-size);
      line-height: var(--xlg-line-height);
    }
  }
`;

export const SummaryTextContainer = styled.div`
  background-color: var(--blue-gray-10-color);
  text-align: center;
  min-height: 80px;
  padding-top: 8px;

  p {
    font-size: var(--xxxs-font-size);
    line-height: var(--xxxs-line-height);
    font-weight: var(--light-font-weight);
    color: var(--blue-gray-900-color);
    margin: 0;
    padding: 24px 32px;
  }

  @media (min-width: ${SMALL_DEVICE_BREAKPOINT}px) {
    text-align: left;
  }
`;
