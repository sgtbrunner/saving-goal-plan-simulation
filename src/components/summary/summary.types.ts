export type FormatCurrencyType = {
  value: number;
  digits?: number;
};
