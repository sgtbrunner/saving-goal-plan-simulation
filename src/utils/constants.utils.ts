// OPERATIONAL CONSTANTS
export const FIRST_STEP = 1;
export const LOCALE = 'en';
export const INITIAL_GOAL_AMOUNT = 25000;
export const PREVIOUS_BUTTON_ID = 'previous-month';
export const NEXT_BUTTON_ID = 'next-month';
export const ARROW_RIGHT = 'ArrowRight';
export const ARROW_LEFT = 'ArrowLeft';

// GENERAL CONSTANTS
export const ID = 'id';
export const ORIGIN = 'Origin';
export const LEFT = 'left';
export const RIGHT = 'right';
export const WHITE = 'white';
export const DOLLAR_SIGN = '$';
