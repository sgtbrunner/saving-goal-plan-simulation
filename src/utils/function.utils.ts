export const clearInvalidCharacters = (value: string): string =>
  value.replace(/\D|^0+/g, '');

export const convertInputToNumber = (value: string): number => {
  if (value.includes('.')) {
    const [decimalPart, fractionalPart] = value.split('.');
    return Number(
      [
        clearInvalidCharacters(decimalPart),
        '.',
        clearInvalidCharacters(fractionalPart),
      ].join('')
    );
  }

  return Number(value.replace(/\D|^0+/g, ''));
};
