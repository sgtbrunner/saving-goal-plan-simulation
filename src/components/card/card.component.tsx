import { CARD } from '../../utils/test.utils';
import { CardContainer, CardTitle, CardForm } from './card.styles';
import { CardProps } from './card.types';
import CustomButton from '../buttons/custom-button/custom-button.component';
import CurrencyInput from '../inputs/currency-input/currency-input.component';
import MonthPickerInput from '../inputs/month-picker-input/month-picker-input.component';
import Summary from '../summary/summary.component';

const Card = ({ title, subtitle, icon }: CardProps): JSX.Element => (
  <CardContainer data-testid={CARD}>
    <CardTitle>
      <img src={icon} alt={title}></img>
      <div>
        <h1>{title}</h1>
        <p>{subtitle}</p>
      </div>
    </CardTitle>
    <CardForm
      onSubmit={(event) => {
        event.preventDefault();
      }}
    >
      <CurrencyInput label="Total amount" />
      <MonthPickerInput label="Reach goal by" />
    </CardForm>
    <Summary />
    <CustomButton>Confirm</CustomButton>
  </CardContainer>
);

export default Card;
