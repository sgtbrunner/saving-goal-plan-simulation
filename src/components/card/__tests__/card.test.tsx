import { screen, RenderResult, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {
  AMOUNT,
  BUTTON,
  CURRENCY_INPUT,
  MONTHLY_AMOUNT,
  renderWithRedux,
  SUMMARY_TEXT,
} from '../../../utils/test.utils';
import buyAHouse from '../../../assets/icons/buy-a-house.svg';
import { CardProps } from '../card.types';
import Card from '../card.component';
import { ID, NEXT_BUTTON_ID } from '../../../utils/constants.utils';
import { getMonthPickerState } from '../../../utils/functions.utils';
import { formatUSDCurrency } from '../../summary/summary.component';

const defaultProps: CardProps = {
  subtitle: 'Saving goal',
  title: 'Buy a house',
  icon: buyAHouse,
};

const mountComponent = (newProps?: CardProps): RenderResult =>
  renderWithRedux(<Card {...defaultProps} {...newProps} />);

describe('Card component', () => {
  it('should match snapshot', () => {
    const { container } = mountComponent();

    expect(container).toMatchSnapshot();
  });

  describe('user interaction events', () => {
    let inputField: HTMLElement;
    let nextButton: HTMLElement;

    beforeEach(() => {
      mountComponent();

      inputField = screen.getByTestId(CURRENCY_INPUT);
      expect(inputField).toBeInTheDocument();
      expect(inputField).toHaveProperty(ID, AMOUNT);

      nextButton = screen.getAllByRole(BUTTON)[1];
      expect(nextButton).toBeInTheDocument();
      expect(nextButton).toHaveProperty(ID, NEXT_BUTTON_ID);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it(`should calculate the monthly amount and monthly deposits and update the
        relevant data in the summary component when the input `, () => {
      const value = 15600;
      const steps = 14;
      // CHANGE THE CURRENCY INPUT TO $15,600
      fireEvent.change(inputField, { target: { value: value.toString() } });
      expect(inputField).toHaveValue('15,600');

      // CHANGE THE REACH GOAL MONTH TO 15 MONTHS AHEAD
      Array(steps)
        .fill(true)
        .forEach((repetition) => userEvent.click(nextButton));

      const { month: monthValue, year: yearValue } = getMonthPickerState(
        steps,
        1
      );

      const month = screen.getByText(monthValue);
      const year = screen.getByText(yearValue);

      expect(month).toBeInTheDocument();
      expect(year).toBeInTheDocument();

      // ASSERT FOR SUMMARY UPDATE
      const monthlyAmount = screen.getByTestId(MONTHLY_AMOUNT);
      expect(monthlyAmount).toHaveTextContent(/monthly amount/i);
      expect(monthlyAmount).toHaveTextContent(
        formatUSDCurrency({ value: value / (steps + 1) }).toString()
      );

      const summaryText = screen.getByTestId(SUMMARY_TEXT);
      expect(summaryText).toHaveTextContent(
        `Youâ€™re planning ${
          steps + 1
        } monthly deposits to reach your ${formatUSDCurrency({
          value,
        })} goal by ${monthValue} ${yearValue}.`
      );
    });
  });
});
