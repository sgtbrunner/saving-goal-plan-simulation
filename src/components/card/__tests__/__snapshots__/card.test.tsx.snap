// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should match snapshot 1`] = `
<div>
  <div
    class="sc-dlnjwi cXHgAV"
    data-testid="card"
  >
    <div
      class="sc-hKFxyN jyhEIY"
    >
      <img
        alt="Buy a house"
        src="buy-a-house.svg"
      />
      <div>
        <h1>
          Buy a house
        </h1>
        <p>
          Saving goal
        </p>
      </div>
    </div>
    <form
      class="sc-eCApnc kGEFpN"
    >
      <div>
        <label
          class="sc-bdnxRM eJleSl"
          for="amount"
        >
          Total amount
        </label>
        <div
          class="sc-gtsrHT eNnCka"
        >
          <div
            class="sc-gKAaRy gPGJKk"
          >
            <span>
              $
            </span>
          </div>
          <input
            class="sc-iCoGMd deleVv"
            data-testid="currency-input"
            id="amount"
            inputmode="decimal"
            name="amount"
            type="text"
            value="25,000"
          />
        </div>
      </div>
      <div>
        <label
          class="sc-bdnxRM eJleSl"
          for="reachDate"
        >
          Reach goal by
        </label>
        <div
          class="sc-gtsrHT sc-pNWdM eNnCka iJUAui"
        >
          <button
            class="sc-fujyAs cXTnUf"
            disabled=""
            id="previous-month"
            type="button"
          >
            <img
              alt="arrow-left"
              src="arrow-left.svg"
            />
          </button>
          <div
            class="sc-jrsJWt hfPpnP"
          >
            <p
              class="sc-kEqXSa gehudG"
            >
              December
            </p>
            <p
              class="sc-iqAclL bfVAAp"
            >
              2021
            </p>
          </div>
          <button
            class="sc-fujyAs cXTnUf"
            id="next-month"
            type="button"
          >
            <img
              alt="arrow-right"
              src="arrow-right.svg"
            />
          </button>
        </div>
      </div>
    </form>
    <div
      class="sc-crzoAE fDjGqz"
    >
      <div
        class="sc-dIsUp dNlBJB"
        data-testid="monthly-amount"
      >
        <h4>
          Monthly amount
        </h4>
        <p>
          $25,000
        </p>
      </div>
      <div
        class="sc-bqGGPW bbfGnO"
        data-testid="summary-text"
      >
        <p>
          Youâ€™re planning 
          <strong>
            1 monthly deposits
          </strong>
           
          to reach your
           
          <strong>
            $25,000
          </strong>
           goal by
           
          <strong>
            December
             
            2021
          </strong>
          .
        </p>
      </div>
    </div>
    <button
      class="sc-jSFjdj kGmRb"
      color="primary"
      type="submit"
    >
      Confirm
    </button>
  </div>
</div>
`;
